ingeniamotion/pdo.py

```diff
@@ -18,6 +18,134 @@
+            buffer_size: Maximum number of register readings to store.
+
+        """
+        super().__init__()
+        self.__mc = mc
+        self.__servo = servo
+        self.__refresh_time = refresh_time
+        self.__buffer_size = buffer_size
+        self.__num_channels: int = 0
+        self.__buffer: List[List[Union[int, float]]] = [[]]
+        self.__timestamps: List[float] = []
+        self.__start_time: Optional[float] = None
+        self.__tpdo_map: TPDOMap = self.__mc.capture.pdo.create_empty_tpdo_map()
+        self.__rpdo_map: RPDOMap = self.__mc.capture.pdo.create_empty_rpdo_map()
+        self.__fill_rpdo_map()
+
+    def start(self) -> None:
+        """Start the poller"""
+        self._clear_buffers()
+        self.__mc.capture.pdo.set_pdo_maps_to_slave(
+            self.__rpdo_map, self.__tpdo_map, servo=self.__servo
+        )
+        self.__mc.capture.pdo.subscribe_to_receive_process_data(self._new_data_available)
+        self.__start_time = time.time()
+        self.__mc.capture.pdo.start_pdos(refresh_rate=self.__refresh_time)
+
+    def stop(self) -> None:
+        """Stop the poller"""
+        self.__mc.capture.pdo.stop_pdos()
+        self.__mc.capture.pdo.unsubscribe_to_receive_process_data(self._new_data_available)
+        self.__mc.capture.pdo.remove_rpdo_map(self.__servo, self.__rpdo_map)
+        self.__mc.capture.pdo.remove_tpdo_map(self.__servo, self.__tpdo_map)
+
+    @property
+    def data(self) -> Tuple[List[float], List[List[Union[int, float]]]]:
+        """
+        Get the poller data. After the data is retrieved, the data buffers are cleared.
+
+        Returns:
+            A list with the readings timestamps and values.
+
+        """
+        data = self.__buffer
+        time_stamps = self.__timestamps
+        self._clear_buffers()
+        return time_stamps, data
+
+    def add_channels(self, registers: List[Dict[str, Union[int, str]]]) -> None:
+        """
+        Configure the PDOs with the registers to be read.
+
+        Args:
+            registers : list of registers to add to the Poller.
+
+        """
+        self.__num_channels = len(registers)
+        self.__fill_tpdo_map(registers)
+
+    def _new_data_available(self) -> None:
+        """Add readings to the buffers.
+
+        Raises:
+            ValueError: If the poller has not been started yet.
+
+        """
+        if len(self.__timestamps) == self.__buffer_size:
+            self.__timestamps.pop(0)
+        if self.__start_time is None:
+            raise ValueError("The poller has not been started yet.")
+        time_stamp = round(time.time() - self.__start_time, 6)
+        self.__timestamps.append(time_stamp)
+        for tpdo_index, tpdo_map_item in enumerate(self.__tpdo_map.items):
+            if len(self.__buffer[tpdo_index]) == self.__buffer_size:
+                self.__buffer[tpdo_index].pop(0)
+            self.__buffer[tpdo_index].append(tpdo_map_item.value)
```

Comment:
Since the buffer size is fixed, we could have better performance by defining the buffer list in the constructor instead of having lists that change dynamically. Then we could have a cyclic variable to store the number of samples so that when the data is retrieved we can return `data[:number_of_samples]`. I think that reducing the time of this callback is wise so we can achieve the best performance of PDOs. What do you think?

Reply:
I'll use a [deque](https://docs.python.org/2/library/collections.html#collections.deque)